numbers/Monte_Carlo.go

```diff
@@ -1,23 +1,80 @@
 package numbers
 
 import (
+	"github.com/vertgenlab/gonomics/common"
 	"log"
 	"math"
 	"math/rand"
+	"time"
+	//DEBUG: "fmt"
 )
 
 //Sample InverseNormal returns a simulated value from a normal distribution.
 func SampleInverseNormal(mu float64, sigma float64) float64 {
 	return rand.NormFloat64()*sigma + mu
 }
 
+//FastRejectionSample returns simulated values from an a func(float64) float64 between a left and right value using an optimized rejection sampler
+//that divides the function support into discrete bins with optimized sampling heights.
+//maxSampleDepth triggers the log.Fatalf in the RejectionSample func, and samples is the number of values to be returned.
+func FastRejectionSampler(xLeft float64, xRight float64, f func(float64) float64, bins int, maxSampleDepth int, samples int, randSeed bool) []float64 {
+	if randSeed {
+		rand.Seed(time.Now().UnixNano())
+	}
+
+	var answer []float64 = make([]float64, samples)
+	if xLeft >= xRight {
+		log.Fatalf("Error in FastRejectionSample: xRight must be greater than xLeft.")
+	}
+	
+	var support float64 = xRight -xLeft
+	var stepSize float64 = support / float64(bins)
+	var binHeights []float64 = make([]float64, bins)
+	var currLeft float64 = xLeft
+	var currRight float64 = xLeft + stepSize
+	var sumHeights float64
+
+	for i := 0; i < bins; i++ {
+		binHeights[i] = common.MaxFloat64(f(currLeft), f(currRight))
+		currLeft += stepSize
+		currRight += stepSize
+		sumHeights += binHeights[i]
+	}
+	var currBin int
+	for j := 0; j < samples; j++ {
+		currBin = chooseBin(sumHeights, binHeights)
```

Comment:
I think chooseBin needs to be inside of the rejection sampling so that it gets chosen again if the x,y pair is rejected.  It should be that b,x,y is chosen and if it fails, new values for all three are chosen.